pinot-spi/src/main/java/org/apache/pinot/spi/data/MultiValueVisitor.java

```diff
@@ -0,0 +1,99 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.spi.data;
+
+import java.math.BigDecimal;
+
+
+public interface MultiValueVisitor<R> {
+
+  R visitInt(int[] value);
```

Comment:
Based on my understanding of visitor, and also from this [blog](https://www.baeldung.com/java-visitor-pattern), visitor should not return anything. Currently we force the visitor to return the same type result when visiting different data types which seems kind of strange. Should we remove the return and always handle the visiting logic within the method?

Reply:
IMO many visitor makes return type generic and we put `Void` type when no return type is necessary. this should be fine.

Reply:
Yep, usually is better to return something in visitors. Otherwise _visitors that return_ should store some state that the caller has to consult later with a getter method. That means that the visitor must be stateful and the code will be uglier. Instead, if you let visitor to return, it will be easier to use from caller perspective and you can always use Void when not needed.

In fact, in a previous life when I heavily use visitors, I usually introduce a second generic `A` (from Argument) that was used to pass contextual information to the visitor. Visit methods were defined as `R visit(SomeClass value, A argument)`. By doing that visitor instances are almost always stateless. I didn't do that here because it is not a well common pattern and didn't want to make it too complex.

Reply:
there's also a visitor return + context
- visitor return carries visiting info (related to the visit sequence) and 
- the context carries contextual info that's unrelated to the visiting sequence

But largely depending on the use case. i don't think we need to go that far here. (we went for this approach in query planner though)